apis/database/v1beta1/sql_types.go

```diff
@@ -162,7 +162,14 @@ type SQLServerParameters struct {
 
 	// TODO(hasheddan): support PublicNetworkAccess
 
-	// TODO(hasheddan): support CreateMode
+	// CreateMode - Possible values include: 'CreateModeDefault', 'CreateModePointInTimeRestore', 'CreateModeGeoRestore', 'CreateModeReplica'
+	CreateMode CreateMode `json:"createMode,omitempty"`
+
+	// RestorePointInTime - Restore point creation time (RFC3339 format), specifying the time to restore from.
+	RestorePointInTime *metav1.Time `json:"restorePointInTime,omitempty"`
+
+	// SourceServerID - The server to restore from when restoring or creating replicas
+	SourceServerID *string `json:"sourceServerID,omitempty"`
```

Comment:
Another optional for `CreateMode` might be to make it a required field, but have it default (using the `// +kubebuilder:default` comment marker) to `CreateModeDefault`.

Reply:
We tested the minimal Tls behaviour with this patch applied as well, from a unified branch, so we know default creates still work.

The zero value for string is "" rather than nil, which just makes handling of the empty case segfault-safe - a nice property to have - and this also lines up with the Azure SDK better. Of course, it is your code base and I can change it, but it will be more code, less ergonomic on the go side, and won't change the Kubernetes API side one iota.

Re: optional - thats only needed when fields have been marked Required, and nothing in this package has been tagged as required to kubebuilder - all the fields are currently optional as far as I can tell - and I've manually verified that this is the case by testing with and without these fields.

Probably we want to fix that, but I think we should do that in one hit, rather than having a mix of correctly annotated and not-annotated fields, because that will be harder to reason about.

Reply:
> We tested the minimal Tls behaviour with this patch applied as well, from a unified branch, so we know default creates still work.

Nice - thank you.

> Re: optional - thats only needed when fields have been marked Required, and nothing in this package has been tagged as required to kubebuilder - all the fields are currently optional as far as I can tell - and I've manually verified that this is the case by testing with and without these fields.

Are we both talking about the API conventions around [optional vs required fields](https://github.com/kubernetes/community/blob/81ece459498964d18159d00948bd4edb849bc8c1/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required)? These state that all optional fields should be pointer types with the `omitempty` JSON struct tag and the `// +optional` comment marker. As far as I can tell this file currently follows those conventions. In this case I believe controller-tools is able to infer that the fields are optional from the `omitempty` tags alone, but I'd like to include the comment markers too for consistency.

Reply:
We weren't. 

On +optional : the rest of the file definitely doesn't follow that consistently, though I can see its present in some places. I'll add it to this and ping when I've done so.

On pointer: I understand the theory of telling 'not supplied string' vs 'empty string' apart, but the Azure SDK we're using doesn't distinguish - an unsupplied string (nil) needs to be converted to an empty string by ForProvider. And in `IsPostgreSQLUpToDate` similarly we'll need to compare a 'nil or ""' string on the Crossplane side to a "" string on the Azure side, always - so I'd really rather not do that, since we cannot use that distinction anywhere. Is that ok?

Reply:
> And in IsPostgreSQLUpToDate similarly we'll need to compare a 'nil or ""' string on the Crossplane side to a "" string on the Azure side, always - so I'd really rather not do that, since we cannot use that distinction anywhere. Is that ok?

No, sorry. I understand where you're coming from with regard to it being technically fairly pointless but I'd like us to comply with the API conventions of the ecosystem we're living in. Having as much consistency as possible across our providers helps us when we get into things generating these types automatically.